### ISSUE TITLE
Unmarshaling always produces zero value?

### ISSUE THREAD

**Issue by ns-cweber** [go-yaml/yaml#412]
I'm assuming I'm doing something wrong, but when I run the following example, I'm always getting the zero value for the `payload`:

Actual output:

```json
{
    "Title": "",
    "Date": "0001-01-01T00:00:00Z",
    "Author": ""
}
```

Desired output:
```json
{
    "Title": "Hello World",
    "Date": "2018-10-31T00:00:00Z",
    "Author": "me"
}
```

Example:

```go
package main

import (
	"encoding/json"
	"log"
	"time"

	"gopkg.in/yaml.v2"
)

const data = `
Title: Hello World
Date: 2018-10-31T00:00:00Z
Author: me
`

func main() {
	var payload struct {
		Title  string
		Date   time.Time
		Author string
	}
	if err := yaml.Unmarshal([]byte(data), &payload); err != nil {
		log.Fatal(err)
	}

	data, err := json.MarshalIndent(payload, "", "    ")
	if err != nil {
		log.Fatal(err)
	}

	println(string(data))
}
```

**Comment by kshitij10496** [comment]
**tldr**: _either add yaml tags to the struct fields or convert the YAML data to have lowercase fields._

The reason this is not working is probably due to the missing YAML tags in the struct type of `payload`. Essentially, the YAML processor searches for _lower-case_ fields when we do not specify the expected field names in the tags. In the above case, the struct field `Title` is mapped to an expected `title` field in the YAML data. I hope you get the drift of why this is not working. If not, let me know and I can elaborate further. 😄 

Fix:

```go

package main

import (
	"encoding/json"
	"log"
	"time"

	"gopkg.in/yaml.v2"
)

const data = `
Title: Hello World
Date: 2018-10-31T00:00:00Z
Author: me
`

func main() {
	var payload struct {
		Title  string    `yaml:"Title"`
		Date   time.Time `yaml:"Date"`
		Author string    `yaml:"Author"`
	}
	if err := yaml.Unmarshal([]byte(data), &payload); err != nil {
		log.Fatal(err)
	}

	data, err := json.MarshalIndent(payload, "", "    ")
	if err != nil {
		log.Fatal(err)
	}

	println(string(data))
}
```

**Comment by ns-cweber** [comment]
Yeah, that was the issue. It expects lower-case fields by default, apparently. I guess I expected it to behave like encoding/json.

**Comment by kshitij10496** [comment]
I think this behavior is consistent with that of `encoding/json` library, isn't it? 🤔 
Can you specify how both of them differ?

**Comment by niemeyer** [comment]
It is a bit different.. go-yaml will encode/decode with the lowercased fields, period. That decision was based on the fact that the vast majority of yaml files I've ever seen are lower-cased. The json package encodes with the original case by default, and relatively recently learned to decode with non-matching casing as a fallback.

Either way, thanks for your comments @kshitij10496. You've nailed it down.
